### PR TITLE
Make `CandidateHash` a real type

### PR DESCRIPTION
**Pull request by bkchr**
This pr adds a new type `CandidateHash` that is used instead of the
opaque `Hash` type. This helps to ensure on the type system level that
we are passing the correct types.

This pr also fixes wrong usage of `relay_parent` as `candidate_hash`
when communicating with the av storage.

### CODE REVIEW

File: node/network/availability-distribution/src/lib.rs [hunk 1]
```diff
@@ -296,8 +296,9 @@ impl ProtocolState {
 				// remove from the ancestry index
 				self.ancestry.remove(relay_parent);
 				// and also remove the actual receipt
-				self.receipts.remove(relay_parent);
-				self.per_candidate.remove(relay_parent);
+				if let Some(candidates) = self.receipts.remove(relay_parent) {
+					candidates.into_iter().for_each(|c| { self.per_candidate.remove(&c.0); });
```
Comment (bkchr) [inline node/network/availability-distribution/src/lib.rs:300]: I think this is actually a bug, as a pending candidate could exists for multiple relay parents? Maybe I'm wrong, I didn't looked closer for now. I just tried to replicate the old behavior.

File: node/network/availability-distribution/src/lib.rs [hunk 2]
```diff
@@ -313,8 +314,9 @@ impl ProtocolState {
 						// remove from the ancestry index
 						self.ancestry.remove(&ancestor);
 						// and also remove the actual receipt
-						self.receipts.remove(&ancestor);
-						self.per_candidate.remove(&ancestor);
+						if let Some(candidates) = self.receipts.remove(&ancestor) {
+							candidates.into_iter().for_each(|c| { self.per_candidate.remove(&c.0); });
```
Comment (bkchr) [inline node/network/availability-distribution/src/lib.rs:318]: Same

File: core-primitives/src/lib.rs [hunk 3]
```diff
@@ -50,6 +50,12 @@ pub type ChainId = u32;
 /// A hash of some data used by the relay chain.
 pub type Hash = sp_core::H256;
 
+/// Unit type wrapper around [`Hash`] that represents a candidate hash.
+///
+/// This type makes it easy to enforce that a hash is a candidate hash on the type level.
```
Comment (coriolinus) [inline core-primitives/src/lib.rs:55]:
```suggestion
/// Unit type wrapper around [`Hash`] that represents a candidate hash.
///
/// This type is produced by [`CandidateReceipt::hash`].
///
/// This type makes it easy to enforce that a hash is a candidate hash on the type level.
```

File: node/network/availability-distribution/src/lib.rs (hunk 1 above)
Comment (montekki) [reply node/network/availability-distribution/src/lib.rs:300]: `CandidateReceipt` includes the `CandidateDescriptor` field that in turn references a `relay_parent`. So two candidates with multiple relay parents will not hash to the same value

**Review by montekki — APPROVED** [review]

File: node/network/availability-distribution/src/lib.rs (hunk 1 above)
Comment (drahnr) [reply node/network/availability-distribution/src/lib.rs:300]: `if descendants.is_empty() {` should prevent this, but given your changes have to look into it again.

**Review by coriolinus — APPROVED** [review]
Looks great! I have not audited the rest of the codebase to ensure that there are no further instances of `Hash` which might correctly become `CandidateHash`. I did look at all instances within this PR of that replacement, and agree that they seem appropriate.

File: node/core/av-store/src/lib.rs [hunk 4]
```diff
@@ -729,7 +729,7 @@ where
 	Ok(())
 }
 
-fn available_data(db: &Arc<dyn KeyValueDB>, candidate_hash: &Hash) -> Option<StoredAvailableData> {
+fn available_data(db: &Arc<dyn KeyValueDB>, candidate_hash: &CandidateHash) -> Option<StoredAvailableData> {
```
Comment (pepyakin) [inline node/core/av-store/src/lib.rs:732]: Wrap this line?

File: node/network/availability-distribution/src/lib.rs [hunk 5]
```diff
@@ -969,7 +970,7 @@ where
 }
 
 /// Query the proof of validity for a particular candidate hash.
-async fn query_data_availability<Context>(ctx: &mut Context, candidate_hash: Hash) -> Result<bool>
+async fn query_data_availability<Context>(ctx: &mut Context, candidate_hash: CandidateHash) -> Result<bool>
```
Comment (pepyakin) [inline node/network/availability-distribution/src/lib.rs:973]: Wrap this line?

File: node/core/av-store/src/lib.rs [hunk 6]
```diff
@@ -564,7 +564,7 @@ where
 				log::trace!(
 					target: LOG_TARGET,
 					"Updating pruning record for finalized block {}",
-					record.candidate_hash,
+					record.block_number,
```
Comment (drahnr) [inline node/core/av-store/src/lib.rs:567]: I guess  this eases reading the output?

**Review by pepyakin — COMMENTED** [review]
some pesky line width nits.

👍 for the newtype

**Review by drahnr — APPROVED** [review]

File: node/core/av-store/src/lib.rs (hunk 6 above)
Comment (bkchr) [reply node/core/av-store/src/lib.rs:567]: It says "finalized_block" and the candidate hash was never and will never be a finalized block :P

File: statement-table/src/lib.rs [hunk 7]
```diff
@@ -18,77 +18,35 @@ pub mod generic;
 
 pub use generic::{Table, Context};
 
-/// Concrete instantiations suitable for v0 primitives.
-pub mod v0 {
```
Comment (rphmeier) [inline statement-table/src/lib.rs:22]: 🙇

**Review by rphmeier — APPROVED** [review]

**Pull request merged**